cpp/src/arrow/dataset/scanner.cc

```diff
@@ -85,6 +92,15 @@ class ScannerRecordBatchReader : public RecordBatchReader {
     return Status::OK();
   }
 
+  Status Close() override {
+    std::shared_ptr<RecordBatch> batch;
+    RETURN_NOT_OK(ReadNext(&batch));
+    while (batch != nullptr) {
+      RETURN_NOT_OK(ReadNext(&batch));
+    }
+    return Status::OK();
```

Comment:
I wonder if this should just be made the default implementation?

Reply:
That seems a little heavyweight.  I don't most of our record batch readers can safely abandon a read partway through.

Reply:
Fair enough.